### PR TITLE
feat(contract-e2e): auto-source substrate-contracts-node with e2e tests

### PR DESCRIPTION
**Pull request by peterwht**
Solves https://github.com/r0gue-io/pop-cli/issues/247

- `pop test contract --features e2e-tests` is now deprecated in favor of `pop test contract --e2e`.
- `pop test contract --e2e` will now download (if needed) `substrate-contracts-node` for usage in e2e tests.
    - If the `substrate-contracts-node` standalone command exists, this will be used.
    - Otherwise, the binary will be downloaded to the pop cache.
- `pop up contract` was refactored to use the existing functionality for sourcing binaries in `pop-parachains`. 

Merge Prerequisite: https://github.com/r0gue-io/pop-cli/pull/258

Review this PR first before #258. #258 moves the sourcing functionality to `pop-common`.

### CODE REVIEW

File: crates/pop-contracts/src/test.rs
Comment (peterwht) [inline crates/pop-contracts/src/test.rs:73]: Needs to be refactored to remove duplicate code

File: crates/pop-contracts/src/test.rs
Comment (peterwht) [inline crates/pop-contracts/src/test.rs:99]: needs refactoring

File: crates/pop-contracts/src/test.rs
Comment (peterwht) [inline crates/pop-contracts/src/test.rs:107]: needs refactoring

File: crates/pop-contracts/src/test.rs
Comment (peterwht) [inline crates/pop-contracts/src/test.rs:125]: needs refactoring

File: crates/pop-cli/src/commands/test/contract.rs
Comment (evilrobot-01) [inline crates/pop-cli/src/commands/test/contract.rs:20]: Missing doc comment, no information available with --help

File: crates/pop-cli/src/commands/test/contract.rs
Comment (evilrobot-01) [inline crates/pop-cli/src/commands/test/contract.rs:51]: format! unnecessary

File: crates/pop-contracts/src/utils/contracts_node.rs
Comment (evilrobot-01) [inline crates/pop-contracts/src/utils/contracts_node.rs:45]: I dont think we should output the version to the console, makes it messy imo.  Can redirect std_err as done elsewhere to silence this. Ideally the version should be read into a variable and returned via a tuple, so that it can potentially be used or output to the terminal in a controlled way.

File: crates/pop-cli/src/commands/test/contract.rs
Comment (evilrobot-01) [inline crates/pop-cli/src/commands/test/contract.rs:50]: Note that none of this triggered for me as I had s-c-n previously installed via cargo. Doing so resulted in a poor experience, so perhaps we only check the cache for the binary for a better experience? It downloads in seconds so think it better to only support pop-managed binaries.

File: crates/pop-contracts/src/utils/contracts_node.rs
Comment (evilrobot-01) [inline crates/pop-contracts/src/utils/contracts_node.rs:56]: Note: this currently doesnt version the binary in the cache, as is done with all the others. Whilst this doesnt need to be addressed in this PR, it should align with the other implementations. The main reason is so that we can check to see if a new version is available and prompt the user to download, improving the devex.

@AlexD10S can we align on binary versioning for the contracts-node please?

File: crates/pop-contracts/src/utils/contracts_node.rs
Comment (evilrobot-01) [inline crates/pop-contracts/src/utils/contracts_node.rs:90]: This seems weird - was this perhaps left in by mistake? Elsewhere we ask for permission to download, but here we just do it anyway.

File: crates/pop-cli/src/commands/test/contract.rs
Comment (evilrobot-01) [inline crates/pop-cli/src/commands/test/contract.rs:57]:
```suggestion
					warning("🚫 substrate-contracts-node is necessary to run e2e tests. Will run tests anyway...")?;
```

**Review by evilrobot-01 — CHANGES_REQUESTED** [review]
Looks great! A few minor suggestions, but need the missing help comment added and I had a weird quick having contracts-node already installed via cargo which meant a poorer experience. See comments for more details.

File: crates/pop-cli/src/commands/test/contract.rs
Comment (peterwht) [reply crates/pop-cli/src/commands/test/contract.rs:51]: We had a discussion about this. We will leave in place for now as it can be nice to use the standalone binary if it exists.

However, there should be some improvements in a follow-up PR, including checking the version and downloading to the pop cache if out-of-date. cc @AlexD10S.

File: crates/pop-contracts/src/utils/contracts_node.rs
Comment (peterwht) [reply crates/pop-contracts/src/utils/contracts_node.rs]: resolved. Please see latest commit and comment.

**Comment by peterwht** [comment]
Introduced a new `common` module under `pop-cli`. A helper function was defined that checks if contracts-node exists, if not, it will prompt to download it.

File: crates/pop-cli/src/commands/test/contract.rs
Comment (AlexD10S) [reply crates/pop-cli/src/commands/test/contract.rs]: With the changes in `does_contracts_node_exist` this is outdated now right? Test it with s-c-n installed via cargo and UX looks good.

File: crates/pop-contracts/src/utils/contracts_node.rs
Comment (AlexD10S) [inline crates/pop-contracts/src/utils/contracts_node.rs:43]: Just a refactor comment here: using a struct might make this clearer than using a tuple (path and version), which would likely eliminate the need for all the comments explaining the return values.

**Comment by AlexD10S** [comment]
Looking good, will review again with this merged: https://github.com/r0gue-io/pop-cli/pull/258 before approve

File: crates/pop-contracts/src/utils/contracts_node.rs
Comment (AlexD10S) [reply crates/pop-contracts/src/utils/contracts_node.rs:58]: Created task

File: crates/pop-cli/src/commands/test/contract.rs
Comment (peterwht) [reply crates/pop-cli/src/commands/test/contract.rs]: For @evilrobot-01, it did something weird where it starting building the binary again. For me, I had no problems with the standalone s c n binary from cargo.

**Comment by AlexD10S** [comment]
> Looking good, will review again with this merged: #258 before approve

Something wrong after the merge, that tests are failing

File: crates/pop-cli/src/commands/test/contract.rs
Comment (AlexD10S) [reply crates/pop-cli/src/commands/test/contract.rs]: After working on the versioning PR: https://github.com/r0gue-io/pop-cli/pull/262 I am in favor of the first comment to only support pop-managed binaries. Supporting other versions of `substrate-contracts-node` makes versioning very hacky.

**Review by AlexD10S — APPROVED** [review]
Tests are passing now.
Some decisions need to be made before merging:
- Can be addressed in a separate PR, but we need to decide on this comment: https://github.com/r0gue-io/pop-cli/pull/254#discussion_r1685341192,  
- Do you want to merge this PR here https://github.com/r0gue-io/pop-cli/pull/262 to have all the functionality together? or separate in another PR to main?.

**Comment by peterwht** [comment]
> Tests are passing now. Some decisions need to be made before merging:
> 
> * Can be addressed in a separate PR, but we need to decide on this comment: [feat(contract-e2e): auto-source substrate-contracts-node with e2e tests #254 (comment)](https://github.com/r0gue-io/pop-cli/pull/254#discussion_r1685341192),
> * Do you want to merge this PR here [refactor: consistency with pop up parachains to recover contracts-node with latest version #262](https://github.com/r0gue-io/pop-cli/pull/262) to have all the functionality together? or separate in another PR to main?.

1. I leave final decision up to you, Alex. But I think just using the Pop cache binary seems like a reasonable approach.
2. I consider PR #262 a feature (not a refactor) as it adds new functionality. So, I would say merge this one, then merge the next. However, I am good with either way.

File: crates/pop-cli/src/commands/test/contract.rs
Comment (AlexD10S) [reply crates/pop-cli/src/commands/test/contract.rs]: Will be tackled in another PR

File: crates/pop-common/src/helpers.rs
Comment (evilrobot-01) [inline crates/pop-common/src/helpers.rs:44]: Personal view.  I dont like `helpers` and `utils`, especially when the root module is empty, as it seems lazy and can potentially make stuff hard to find for users of the crate. From a crate consumer perspective, would it not be easier to find `target` within `pop_common`? From a scope perspective, does `target` not potentially apply to the whole crate in terms of usage. In other words, structuring things hierarchically from general to specific.

If we dont want it in the root, then would something like `platform` not be more descriptive/useful then `helpers`?

Only mentioning it as it starts out as ill-considered and then everybody else just follows the same approach and then we end up with a module with a whole lot of random functions that dont have any real connection.

File: crates/pop-contracts/src/call.rs
Comment (evilrobot-01) [inline crates/pop-contracts/src/call.rs:315]: As above, use https://doc.rust-lang.org/std/env/fn.temp_dir.html if temp_dir doesnt work, so that stuff is at least auto-cleaned on restart. Currently the remove_dir_all will not run if there is an error in any of the below lines using ?

File: crates/pop-contracts/src/up.rs
Comment (evilrobot-01) [inline crates/pop-contracts/src/up.rs:343]: Same as above.

File: crates/pop-contracts/src/utils/contracts_node.rs
Comment (evilrobot-01) [inline crates/pop-contracts/src/utils/contracts_node.rs:45]: As above, why 'utils' and not just a 'node' module in the root of the crate? A node is integral to using/testing/deploying contracts and should be a first-class citizen, not stuffed away under utils.

File: crates/pop-contracts/src/utils/contracts_node.rs
Comment (evilrobot-01) [inline crates/pop-contracts/src/utils/contracts_node.rs:160]: As above with using cache dir.

File: crates/pop-cli/src/commands/test/contract.rs
Comment (evilrobot-01) [inline crates/pop-cli/src/commands/test/contract.rs:34]: This needs to be fixed. The `intro` should always be first, due to the various bar symbols used by cliclack.

**Review by evilrobot-01 — CHANGES_REQUESTED** [review]
A minor fix required to the UX and the use of the cache directory for tests need to be improved, as it may currently leave artifacts on users machine if test fails.

See also https://github.com/r0gue-io/pop-cli/pull/263 for additional improvements.

File: crates/pop-cli/src/common/contracts.rs
Comment (peterwht) [reply crates/pop-cli/src/common/contracts.rs:7]: This is resolved in: https://github.com/r0gue-io/pop-cli/pull/262

File: crates/pop-cli/tests/contract.rs
Comment (peterwht) [reply crates/pop-cli/tests/contract.rs:47]: It's because the temp_dir handle gets moved in the subsequent functioned calls, so, although the original handle still exists (when using clone), the clones get dropped which result in the directory getting prematurely cleaned.

Passing references introduces unnecessary complexity in terms of manually specifying lifetimes on the `Binary` struct. So, I felt manually creating a temp_dir is the least overhead for a test.

Regarding the leftover tmp, in-practice a failing test is usually re-run until it is fixed. I will continue looking for a better way.

Comment (evilrobot-01) [reply crates/pop-cli/tests/contract.rs:47]: Does temp_dir.path().join() not work?

**Review by evilrobot-01 — APPROVED** [review]
Awesome, thank you for bearing with me. 🙂

**Pull request merged**
